addons/project/models/project_task_type.py

```diff
@@ -79,6 +84,60 @@ def copy(self, default=None):
             default['name'] = _("%s (copy)") % (self.name)
         return super().copy(default)
 
+    def unlink(self):
+        # 1. Personal stages are processed if the user still has at least one personal stage after unlink
+        personal_stages = self.filtered('user_id')
+        remaining_personal_stages_all = self.env['project.task.type']._read_group(
+            [('user_id', 'in', personal_stages.user_id.ids), ('id', 'not in', personal_stages.ids)],
+            groupby=['user_id', 'sequence', 'id'],
+            order="user_id,sequence DESC",
+        )
+        remaining_personal_stages_by_user = defaultdict(list)
+        for user, sequence, stage in remaining_personal_stages_all:
+            remaining_personal_stages_by_user[user].append({'id': stage.id, 'seq': sequence})
+
+        # For performance issue, project.task.stage.personal records that need to be modified are listed before calling _prepare_personal_stages_deletion
+        personal_stages_to_update = self.env['project.task.stage.personal']._read_group([('stage_id', 'in', personal_stages.ids)], ['stage_id'], ['id:recordset'])
+        for user in personal_stages.user_id:
+            user_stages_to_unlink = personal_stages.filtered(lambda stage: stage.user_id == user)
+            user_remaining_stages = remaining_personal_stages_by_user[user]
+            if not user_remaining_stages:
+                raise UserError(_("Each user should have at least one personal stage. Create a new stage to which the tasks can be transferred after the selected ones are deleted."))
```

Comment:
We could avoid throwing that error if the user is archived (or if the user is no longer an internal user). 🙂 

It would be useful when we will re-introduce the garbage collector to remove the personal stages of archived user. 🙂

Reply:
> We could avoid throwing that error if the user is archived (or if the user is no longer an internal user). 🙂
> 
> It would be useful when we will re-introduce the garbage collector to remove the personal stages of archived user. 🙂

By "no longer an internal user", do you mean no more base.group_user access?

Reply:
Exactly, the user could become a portal user for instance. I agree it is not a common use case but since the user can change the user. 🙂